### PR TITLE
provider: add `assume_role.0.duration` and deprecate `assume_role.0.duration_seconds`

### PR DESCRIPTION
**Pull request by anGie44**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes https://github.com/hashicorp/terraform-provider-aws/issues/23028

Output from acceptance testing:

```
--- PASS: TestValidTimeDuration (0.00s)

--- PASS: TestAccAcctestProvider_Region_awsSC2S (29.78s)
--- PASS: TestAccAcctestProvider_Region_awsC2S (29.79s)
--- PASS: TestAccAcctestProvider_Region_awsCommercial (30.50s)
--- PASS: TestAccAcctestProvider_IgnoreTagsKeys_one (32.20s)
--- PASS: TestAccAcctestProvider_DefaultAndIgnoreTags_emptyBlocks (32.53s)
--- PASS: TestAccAcctestProvider_IgnoreTagsKeys_multiple (32.56s)
--- PASS: TestAccAcctestProvider_IgnoreTagsKeyPrefixes_multiple (32.56s)
--- PASS: TestAccAcctestProvider_IgnoreTags_emptyBlock (32.57s)
--- PASS: TestAccAcctestProvider_DefaultTagsTags_one (32.57s)
--- PASS: TestAccAcctestProvider_IgnoreTagsKeyPrefixes_none (32.60s)
--- PASS: TestAccAcctestProvider_DefaultTags_emptyBlock (32.86s)
--- PASS: TestAccAcctestProvider_DefaultTagsTags_none (32.94s)
--- PASS: TestAccAcctestProvider_IgnoreTagsKeyPrefixes_one (34.12s)
--- PASS: TestAccAcctestProvider_IgnoreTagsKeys_none (34.18s)
--- PASS: TestAccAcctestProvider_DefaultTagsTags_multiple (34.35s)
--- PASS: TestAccAcctestProvider_endpoints (34.64s)
--- PASS: TestAccAcctestProvider_AssumeRole_empty (35.44s)
--- PASS: TestAccAcctestProvider_unusualEndpoints (10.27s)
--- PASS: TestAccAcctestProvider_fipsEndpoint (47.65s)
```

### CODE REVIEW

File: internal/provider/provider.go [hunk 1]
```diff
@@ -1936,11 +1936,20 @@ func assumeRoleSchema() *schema.Schema {
 		MaxItems: 1,
 		Elem: &schema.Resource{
 			Schema: map[string]*schema.Schema{
+				"duration": {
+					Type:          schema.TypeString,
+					Optional:      true,
+					Description:   "The duration of the role session, e.g. 1h. Valid time units are ns, us (or µs), ms, s, h, or m.",
+					ValidateFunc:  verify.ValidTimeDuration,
```
Comment (anGie44) [inline internal/provider/provider.go:1943]: Q for reviewer: This validation does not allow for empty strings...or should it?

File: website/docs/index.html.markdown [hunk 2]
```diff
@@ -335,7 +335,8 @@ In addition to [generic `provider` arguments](https://www.terraform.io/docs/conf
 
 The `assume_role` configuration block supports the following optional arguments:
 
-* `duration_seconds` - (Optional) Number of seconds to restrict the assume role session duration. You can provide a value from 900 seconds (15 minutes) up to the maximum session duration setting for the role.
+* `duration` - (Optional, Conflicts with `duration_seconds`) The assume role duration represented as a string such as `300ms`, `-1.5h` or `2h45m`. Valid time units are `ns`, `us` (or `µs`), `ms`, `s`, `m`, `h`.
```
Comment (gdavison) [inline website/docs/index.html.markdown:338]: We don't accept negative duration and the minimum is 15 minutes. Something like the following (borrowed from AWSCC)

```suggestion
* `duration` - (Optional, Conflicts with `duration_seconds`) Duration of the assume role session. You can provide a value from 15 minutes up to the maximum session duration setting for the role. Represented by a string such as `1h`, `2h45m`, or `30m15s`.
```

File: website/docs/guides/version-4-upgrade.html.md [hunk 3]
```diff
@@ -99,6 +99,7 @@ provider "aws" {
 
 Version 4.0.0 adds these new provider arguments:
 
+* `assume_role.0.duration` - The assume role duration represented as a string e.g. `"1h"` or `"1h30s"`. Replaces `assume_role.0.duration_seconds`, which has been deprecated in Terraform AWS Provider v4.0.0 and support will be removed in a future version.
```
Comment (gdavison) [inline website/docs/guides/version-4-upgrade.html.md:102]: The `.0.` is something internal to the SDK representation, so users don't see it
```suggestion
* `assume_role.duration` - The assume role duration represented as a string e.g. `"1h"` or `"1h30s"`. Replaces `assume_role.duration_seconds`, which has been deprecated in Terraform AWS Provider v4.0.0 and support will be removed in a future version.
```

File: internal/verify/validate.go [hunk 4]
```diff
@@ -257,6 +257,17 @@ func ValidStringIsJSONOrYAML(v interface{}, k string) (ws []string, errors []err
 	return
 }
 
+// ValidTimeDuration validates a string can be parsed as a valid time.Duration
+func ValidTimeDuration(v interface{}, k string) (ws []string, errors []error) {
+	_, err := time.ParseDuration(v.(string))
+
+	if err != nil {
+		errors = append(errors, fmt.Errorf("%q cannot be parsed as a valid time.Duration: %w", k, err))
```
Comment (gdavison) [inline internal/verify/validate.go:265]: Referencing `time.Duration` leaks an internal implementation detail to practitioners. Maybe something like

```suggestion
		errors = append(errors, fmt.Errorf("%q cannot be parsed as a duration: %w", k, err))
```

File: internal/provider/provider.go (hunk 1 above)
Comment (gdavison) [reply internal/provider/provider.go:1943]: Looking at the validation functions provided by the Plugin SDK, and especially [`validation.IsRFC3339Time()`](https://github.com/hashicorp/terraform-plugin-sdk/blob/36ee3aa4bc4b9a4fa75ccf85b481293fa51c1ecc/helper/validation/time.go#L42), we're probably fine not checking for it.

Comment (gdavison) [inline internal/provider/provider.go:1943]: We could also check for the minimum value of 15 minutes and the absolute maximum of 12 hours

**Review by gdavison — COMMENTED** [review]
Looks good. I have a few comments

File: website/docs/index.html.markdown (hunk 2 above)
Comment (anGie44) [reply website/docs/index.html.markdown:338]: sounds good 👍 ooof I should of caught the negative one and looked at the awscc definition

File: internal/provider/provider.go (hunk 1 above)
Comment (anGie44) [reply internal/provider/provider.go]: Test output after updates:
```
--- PASS: TestValidAssumeRoleDuration (0.00s)
```

**Review by gdavison — APPROVED** [review]
Looks good! 🚀

**Pull request merged**